### ISSUE TITLE
Street name and lane marker disappear.

### ISSUE THREAD

**Issue by PaulStets** [osmandapp/Osmand#3898]
The street name and lane marker show up like normal when I start riding, and once I turn off of that first road, they disappear and never come back.

Android 5.1.1, OsmAnd version 2.6.5

Video: https://drive.google.com/file/d/0B_epabEezdJreFV4Ql8wLWd6eEk/view?usp=sharing

**Comment by vshcherb** [comment]
Do we have this part of GPX file?

**Comment by PaulStets** [comment]
I've asked for it. Will post it when the user responds.

**Comment by PaulStets** [comment]
Here it is:
https://www.dropbox.com/s/jx7d5f16vc24dzi/2017-06-03_16-59_Sat.gpx?dl=0

**Comment by polarbearing** [comment]
The video does not load for me. However it sounds like an issue I experienced on my old single core Nexus S phone, see #2962. 
Naturally, when the street is not recognised, there is no lane information. 
Did not yet have the issue on my current multicore device.

**Comment by tazva** [comment]
I published the issue to Paul. It doesn't matter what street it's on, after a short while (usually within a few miles), osmand simply stops displaying the street names and lane indicators. It worked fine for months, then one day simply stopped working--I believe after an upgrade, but cannot recall the specific day.

**Comment by tazva** [comment]
I have the app on another device running android 6.0.1, I've noticed that this bug does not appear on that device, only my tablet which is running 4.4.2.

**Comment by sonora** [comment]
Very weird, I can almost 100% reproduce this on a Samsung S4 Mini with Android 4.4.2 stock ROM, in a simulation. Losing the street name display at almost the identical spots. But it does actually recover, also reproducible, at fixed locations:

- In your video, the first loss happens at 1:31, going from "Butts Station Road " via very briefly showing "Near Butts Station Road" (which seems to indicate OsmAnd somehow moves away from a named way segment) to blank, but it recovers at around 1:44.

- Your second occurrence happens at 3.46 losing "Greenbrier Parkway", again via a brief appearance of "Near Greenbrier Parkway". It does not recover in the time span of your video, but when I simulate this, it actually recovers after pulling onto I-64, when it starts displaying "I 64" about near http://www.openstreetmap.org/way/391802357. Very briefly only, however, as more losses occur thereafter when proceeding down I 64. The next brief appearance of "I 64" is right after the Indian River Road bridge. These losses and re-coveries seem to reliably happen at reproducible points, I think.

It does not seem to be an issue with our reverse geocoding code, because if I manually select a point on the map at a stretch where (and during) the street name widget shows blank, the selected point's context menu correctly resolves and shows the address as e.g. "Greenbrier Parkway" (although that name could be retrieved from e.g. the opposite way of the divided highway. Looks like most or all occurrences happen on divided highways.)

Rather weird, needs more investigation. I would normally suspect a mapping issue, but cannot find one at first sight. Maybe someone more experienced with mapping an take a look at the ways in question and check out if there is anything faulty.

**Comment by tazva** [comment]
I'm an OSM contributor, and have examined various pathways where this occurs. At first I believed it to be because some points lacked road names, but then it would happen in places where there's no lack or break of road names. On my system, it sporadically "recovers", but only briefly. One day last week I was riding somewhere (nowhere near any of the locations in the evidence provided), and suddenly a street name popped up, but a few seconds later it did the "Near ...." thing, and then it dropped off again.

If there's any specific roads I should examine in OSM, let me know. Personally, I don't think it's OSM-related, because it didn't do this prior to the recent upgrade. I'm more inclined to believe it's an implementation/method of code.

I've also begun to notice that in Android 6x, the new version seems to lose track of GPS when walking, something it didn't do before--but that's for another thread.

**Comment by sonora** [comment]
Thanks for investigating. Have you seen any occurrences on non-divided highways?

Regarding loss of GPS on newer Android systems: Chances are this is related to new power saving features, check out e.g. https://github.com/osmandapp/Osmand/issues/3923#issuecomment-309232409 .

**Comment by tazva** [comment]
I see occurrences of this thread's issue everywhere. I haven't come across anything where it hasn't happened. First boot, "Road" shows up. A short while later "Near Road", and finally nothing.

Regarding the GPS loss, I thought about power, but other GPS apps aren't losing it. For example, I stopped using OsmAnd to record GPX on my newer devices because it "can't see" the GPS for long periods of time, whereas my other GPS items (and GPX logger), never lose a connection. I examined my power options, I have all power saving disabled.

**Comment by tazva** [comment]
I actually had a moment of surprise when the functionality remained intact yesterday for about 10 minutes before dropping out. No apparent reason why.

**Comment by tazva** [comment]
This is an odd question, but is it possible the reverse lookup (which I'm presuming is whats being used to query the name for display), is timing out? Or whatever function passes the coordinates is getting backlogged or gets stuck? It seems like nearly every single "loss" is prefixed by the "near... road" display, which almost seems like OSM StreetName gets stuck on a node or street, and as I get further away its next update renders to "Near" that road, and then finally it just stops.

**Comment by sonora** [comment]
Yes, I have a similar suspicion. But strangely, it seems that during such an "outage" I can tap any POI or long-tap any other point on the map and the popup context menu **will actually correctly resolve and display its address**. So it seems that not the reverse geocoding lookup gets overall stuck, but somehow maybe only the lookup process for the current/next street segment lookup.

Not sure if that makes sense, but I think Victor (@vshcherb) needs to take a look at the symptoms we found, he had discovered and fixed similar reverse lookup process collisions in the past.

**Comment by tazva** [comment]
I noticed today that for the brief amount of time it's working, if I'm near
a marker point such as a rail road, the icon will show up in the corner
like it should. It then takes a street change to disappear no matter the
distance away from the rail road. Not sure if this is helpful or not.

I'll try the manual reverse lookup myself since you mentioned it.

**Comment by tazva** [comment]
I tried the reverse lookup as you suggested. I tried this after it had lost the street display. No-dice.. couldn't get it to reverse lookup anything by selecting a point. It shows the coordinates just fine, and then it'll say "looking up address" (or whatever), and that's all it does. I tried a bunch, moved around the map selecting different roads and buildings.. nothing, it just sits there in an endless "looking up address".

I'll try it again before it loses street display.

**Comment by tazva** [comment]
After a fresh boot of the tablet today, I waited until OSM had fully loaded in and the street name was showing. I used the reverse lookup to check various streets with success. After about a minute of moving down the road, the street/everything went out like usual, and reverse lookup stopped functioning.

**Comment by sonora** [comment]
Not my observation, actually. I think the address lookup still works, it may just take a lot longer during an ongoing navigation, probably a matter of CPU performance.

But I am afraid I was heading down the wrong track here anyway, because we do not use reverse geocoding lookup to determine the street name.

**Comment by sonora** [comment]
Code ref to check: https://github.com/osmandapp/Osmand/blob/master/OsmAnd/src/net/osmand/plus/views/mapwidgets/MapInfoWidgetsFactory.java#L737

**Comment by sonora** [comment]
That sounds somehow like a process competition...something somehow taking REALLY long. And that matches my observation. And it seems perfectly reproducible when repeating the exact same steps e.g. by running a simulation along the identical GPX track.

Can you specify with which update exactly you think this started? Thx!

**Comment by tazva** [comment]
Whichever one provided the animation smoothing for your current location.. I want to say it was the most recent update.

Also, if this is helpful, this doesn't occur during navigation (that I can recall, I will test it), but it's when I'm just riding around.

**Comment by sonora** [comment]
When I **simulate** a navigation along the GPX track provided above, it seems to happen on my device at the exact same places as shown in the video (which shows the just-cruising case). I have however, not seen the effect (yet) for not simulated "real" navigation.

**Comment by sonora** [comment]
After lots of more testing, it turns out this is **not 100%** reproducible. I simulated the stretch on Greenbrier Parkway in front of Greenbrier Mall now 12 times. In 10 cases, I lost the "Greenbrier Parkway" Display just after crossing Eden Way North, but in 2 cases I did get the display ok together with the "3 lanes" indicator just fine all the way to the I-64 ramp.

I think we can rule out that this is a simple screen refresh issue because of this additional observation: The CPU consumption my Task Manager reports for OsmAnd even with OsmAnd in the background is really high after occurrences of this issue (about 50%-60% continuously until I kill OsmAnd), whereas normally it drops to 0% when OsmAnd is just idling in the background and no address lookup is ongoing. I normally see high CPU usage for OsmAnd while it is in the background in non-navigation mode only exactly during the periods when an address lookup is till ongoing, then dropping to 0% after an address has been found.

So it really looks like we are dealing with a process competition issue here. **I now tend to think that this issue is an exact duplicate of #2962.**

**Comment by sonora** [comment]
PS: Also related to https://github.com/osmandapp/Osmand/issues/2960#issuecomment-240542172. I guess we have to re-visit how we organize our reverse geocoding lookup, and also investigate to cache results to some extend because repeated attempts all the time are quite CPU/power intensive.

**Comment by tazva** [comment]
Yeah, those threads sound exactly like my issue.

**Comment by vshcherb** [comment]
@sonora geocoding lookup - shouldn't happen during navigation. It uses navigational data which is precalculated. As I saw in the video it was a navigation case.

**Comment by tazva** [comment]
I wasn't navigating anywhere in the video, I was just riding along. If
that's what you meant by navigation.

I just tried osmand on another device of mine running Android 6, quad
2.7ghz device. It happens on that as well, although not as quickly as with
my tablet.

**Comment by sonora** [comment]
Victor, I was confused at first as well, but we are talking the non-navigation case here. Although I could also reproduce this using a **simulated** navigation in Follow mode.

**Comment by vshcherb** [comment]
For non-navigational case that's a known issue, it just jumps similar to how GPS jumps. So until we teach GPS signal to stay tuned or learn more about the history movement, it will be like this.

**Comment by tazva** [comment]
There's got to be a way around it. I find it odd  to believe a dedicated
1.4ghz or 2.7 ghz system can't process lookups quick enough, especially
since it worked fine for me for months with no other changes. It just seems
odd. Lookups could be narrowed down in some fashion, it's not like I have a
country's worth of maps downloaded, just one local map.

It seems to simply get stuck, taking hours to undo itself, sometimes never.
That isn't a simple delay, that's the process  getting lost.

**Comment by sonora** [comment]
What really bothers me is the high CPU load during and after this occurs, this really needs fixing. It seems that OsmAnd keeps looking up the address but never finishes or times out, and this is really a CPU (and eventually battery) eater. I can even feel the difference in temperature of a device in my hand between when exhibiting the issue and when not.

**Comment by sonora** [comment]
I went through all the evidence again, and it looks to me like this: While moving without navigation but with the street name widget on, we **queue** an enormous number of reverse address lookups. Looks like during motion we hardly ever catch up again. That is also why CPU usage remains enormous even when you put OsmAnd in the background then. But when you stop and leave the device at rest for a (potentially very) long time, this eventually does catch up (until you start moving again).

Maybe the solution is as simple as not queuing any lookup processes (a new one always cancels an old one), or we somehow strictly limit the number of requests waiting for resolution? Or we find a way to flag  requests as outdated, like all unresolved ones which were triggered by successive calls from/for the street name widget.

**Comment by tazva** [comment]
Your explanation matches my observation from yesterday. I ran OSM on my 4x2.7ghz device for a short while (10-15 minutes), and in that duration, it became very warm and killed off almost 30% of the battery. It continued to do so long after I'd stopped moving (but the street name never came back), so I had to kill the OSM process. This may also explain why OSM drains so much battery if I leave my tablet on but motionless, and the CPU never idles down.

In addition to the suggestions you've made, perhaps some sort of rate trigger could be included, based on the speed, because the faster you travel the greater the distance from your previous point. Perhaps all of these ideas could be put into a dev/beta menu, and could be toggled on/off to see what helps? I don't mind toggling my beta opt-in on if we can help work this out.

**Comment by scaidermern** [comment]
I've experienced several very long delays when adding OSM notes while moving. The note dialog opens up immediately, but after closing it it takes a very long time until OsmAnd shows the window where you can select to upload or delete the note, sometimes this takes 30 minutes or longer. Maybe this is caused by the same problem.

What is the reason for *queuing* reverse address lookups? Do we need older addresses at all?

**Comment by sonora** [comment]
To queue a small number probably makes sense, like for when you add several interim destinations or map markers in a row, you may want to see the address resolution for all of them, not just the last one. But it makes no sense if we clog the process by dozens of successive resolution requests as you move along segments of OSM ways.

**Comment by vshcherb** [comment]
It searches every 10-50 meters and that could be a problem (we could fix it)
```
if (loc == null || loc.getAccuracy() > 50) {
			return null;
		}
		if(last != null && last.distanceTo(loc) < 10) {
			return r;
		}
		if (r == null) {
			scheduleRouteSegmentFind(loc, true, null, null);
			return null;
		}
		double d = getOrthogonalDistance(r, loc);
		if (d > 15) {
			scheduleRouteSegmentFind(loc, true, null, null);
		}
		if (d < 70) {
			return r;
		}
```

**Comment by tazva** [comment]
I've had issues saving GPX files manually, where sometimes it wouldn't save at all, or it took what seemed like forever. Selecting 'save' just puts the word "SAVE" in the GPX widget, and the file is often not saved, is sometimes lost, and sometimes is saved days later. I'm wondering if the lag issue we're discussing is causing problems for the GPX saving? There's also times GPX doesn't reset to zero, even several days later, it just keeps building.

This guy had a similiar issue with failure to zero out upon saving: https://github.com/osmandapp/Osmand/issues/2479

**Comment by vshcherb** [comment]
The cpu usage is quite high mostly because no caching is use between interactions. Which could be a nice improvement.

**Comment by tazva** [comment]
I'm thinking this overhead issue is what's causing a cascade of other minor issues, such as my GPX file not "resetting" sometimes, and then for no apparent reason, resetting in the middle of a drive.

I also noticed on a couple of occasions where I was in navigation mode, it had to re-route, and re-routing was so slow that at one point, I pulled over and waited for 10 minutes while it continued to 'calculate'. i eventually just had to close the application and re-open it. Sometimes re-routing is instant, sometimes it takes 30 seconds or a minute, just for missing one turn, despite the streets all being parallel.

**Comment by tazva** [comment]
Here's a couple more data points.

I was in navigation mode the other day  for about half an hour, and during this time the street name disappeared a handful of times. When it did, it never came back, but when I crossed onto the next set of directions/new street, that name would pop up. When the name disappeared, it left the directional icon up in the bar (so the left turn arrow for example). It didn't disappear a lot, but it was happening.

I notice that as it switches to night mode, it seems to be a very slow transition, particularly once the street name has disappeared (and we presume OSMA is backlogged). It slowly fills sections of the display with the night theme, sometimes takes 5 seconds, sometimes takes 20 seconds.

I'm presuming all of this to be due to the backlog issue we've discussed here. I'm using the most current version of OSMAnd+.

**Comment by tazva** [comment]
Adding to my comment above, in nav mode, it seemed less likely for the street name to disappear, whereas in non-nav mode, you can count on the name disappearing.

**Comment by tazva** [comment]
While in night-mode, I've noticed that as it transitions between various map levels, this transition can sometimes be 'slow' in drawing on screen. This results in patches of the map being shown in 'day' mode, which causes a 'bright white' flashing on screen (because the light map colors are contrasting fast with the dark map colors).

**Comment by scaidermern** [comment]
I just experienced some of these issues again during hiking. I had the walking profile active, so the street name widget was _not_ enabled. Nevertheless after some hours I did encounter long delays on various actions. E.g. adding an OSM note or editing a POI. The corresponding dialogs did appear only after several minutes.

**Comment by tazva** [comment]
I recently used this to navigate a short distance away from my origination point, it went fairly well. Later in the day another navigation episode was a little more rough, as a few times it had to reroute due to road construction, and at one point it took so long to reroute I simply pulled over and had to cancel it all out and reload everything.

There's been a couple of times this week where its attempt to "catch up" was burning so much energy, the device was unable to "charge" and simply "maintained" its capacity at the same % for the duration.

**Comment by tazva** [comment]
I ran this on a Note 4 Edge over the weekend, to monitor some OSM map changes over the course of a 5 minute trip. During this time, the street display washed in/out (as described in this thread), CPU usage was absolutely off the charts, and I lost 20% of the battery of the phone. This is absolutely ridiculous. 

I would have to put the device into an ARM stress test just to match the resources Osmand+ is using when this happens. On the bright side, if I ever needed to heat my hands up while in Russia during winter, I could just load the app and hold onto the phone.

**Comment by tazva** [comment]
I used this on a Note 4 Edge a week ago to check some OSM edits. I only needed to drive about 5 miles/10 minutes. In this time, the road display cut out a few times for short periods (as described in this thread), the device was unreasonably hot, and I lost 20% of my battery power, all in less than 10 minutes. 2.7ghz quad core Android device having chugging issues.. it really is hard to believe.

This is pretty crazy. If I need a hand warmer in Siberia, I could just load Osmand+ onto a device and hold it with my hands.

Having said that, it doesn't happen all the time on the device--but nothing else was running. It seems sporadic, but when it does get stuck in these "bad loops", it simply never stops.

**Comment by CJTmmr** [comment]
Can confirm excessive CPU-use upto 27%..when not displaying current streetname, lanes when in simple follow-me -mode. Even when OsmAnd is no longer running in foreground. On Samsung Galaxy Camera2, Android 4.3, 4 cores.
When somehow display is okay again CPU is back to normal.
Some search proces gets probably in a endless loop.

**Comment by tazva** [comment]
In the interest of data points, I'll see if I can log cpu usage while using the app. Watching it the other day, cpu usage is 60-90% for all 4 cores on my device, at any given point.

**Comment by tazva** [comment]
A week ago I said I'd start logging load to see if I can provide any further useful objective data. **The purpose of this wasn't to prove this issue (as it has been observed by many), but to examine what it actually looks like, from a data-visual perspective.**

In order to monitor process usage and log CPU load, I used Cool Tool Pro from deviantstudio (https://play.google.com/store/apps/details?id=ds.cooltool). This lets you use a window overlay to monitor chosen fields. 

In my case, for the window overlay, I keep an eye on battery%, cpu load%, memory free, temperature, and most important, what system process is using the most resources. 

**My observations:** 
For visual monitoring, I have it set to 1s update intervals. For logging CPU load to a file, I had it set to 
30s intervals (I forgot to set it to 10s). I've kept an eye on the overlay over the past week, and OSM+ cpu usage was much higher than I realized.. at idle it's burning up the CPU for no apparent reason. By idle, I mean just turning the tablet on and letting it sit motionless. GPS shows no movement, but OSM+ is typically loading the CPU 30-40% at idle.

When not idle (ie, when simply moving on roads), usage typically bounces between 60-80%, with spikes to 100%. Navigation mode doesn't appear to make a difference (good or bad). I've also twiddled with GPX recording, it may make a hair difference (couple of %), but appears to be a victim of this issue more than anything.

To try and provide a simple visual illustration of my findings, I transposed the CPU load log results with GPX recordings of trips, one data group using navigation to get to a destination, and the other was simply riding around town--the data was tied together using system timestamp. GPX interval for nav mode is 3s, while in explore mode is 2s.

Below is a snapshot showing these two situations. The top chart is "cruise" mode, simply exploring, bottom chart is "navigation" mode. Red line is speed in MPH (axis to the right), blue is CPU load in % (axis to the left).

![osmandplus_load](https://user-images.githubusercontent.com/29614385/32289330-d3ec5a8a-bf0c-11e7-9774-16b5eda18528.jpg)

Notice in exploration mode (top), CPU load quickly ramps up to the 70% range and stays there until movement is at a stop, once at a stop it drops to the 30% range. Occasionally while at rest, it will spike--doubling--for no apparent reason. All of this usage is OSM+. The duration for exploration was around an hour--I truncated the speed timestamp of the tail because it was motionless, but allowed the load to be displayed to show that it was using very much CPU despite being at rest doing absolutely nothing.

In navigation mode (bottom), you can see it's not much different. There's thousands more data points here due to the duration (5+ hours), so the chart is thicker. **Again, OSM+ "idle" appears to just use 30% of CPU for no apparent reason (the 4 hour gap of being at rest)**, and its "normal operating load" appears to be 60-70% at any given point (which corresponds to the operating load for non-nav mode).

For anyone curious, Base CPU load (booted but without osm+, doing nothing) is < 4%.

Other data points I examined mirror this, OSM+ load at idle being very high, and while in use, more than doubles. My concern is that moving forward (building on the existing code/bug or whatever is causing this), is going to be hurtful to the app, because CPU usage makes it very inefficient. I believe it's also going to cause other issues, as more features will become available, but performance may be sluggish. The fact that I can install this on a new device, and drain the device battery by the second, is further proof of this. It seems there's an inefficiency in the code somewhere. I've mirrored these results on another device, and it's rate is no different.

When I first got this app, I was excited at the prospect of being able to explore and see the current road and the number of lanes, because I'm an OSM developer, and this helps me test changes in the real world environment. With this inefficiency, I can no longer do that, because these displays disappear, as mentioned in the beginning of this thread.

**Comment by tazva** [comment]
Here's some screenshots showing the user-side of the data collection. I was travelling yesterday as a passenger, and had the chance to capture these. The first one shows you the base performance of the tablet  with everything but OSM+ loaded up.

![screenshot_2017-11-01-14-06-46](https://user-images.githubusercontent.com/29614385/32377356-bb540726-c07d-11e7-91db-eafe25a7eb19.png)

The following are some shots taken a few seconds apart.  Note the lack of street name/lane markers, which disappeared moments into loading OSM+, as the CPU load ramped up.

![screenshot_2017-11-02-13-13-20](https://user-images.githubusercontent.com/29614385/32377357-bb8a71ee-c07d-11e7-897f-5fb628552555.png)

Notice with the load being so high, when OSM+ goes to automatically change zoom levels based on speed, it's unable to clearly render the map for a short amount of time, and the overlays disappear.

![screenshot_2017-11-02-13-13-23](https://user-images.githubusercontent.com/29614385/32377353-bae62026-c07d-11e7-9522-e5f2a8ef4608.png)
![screenshot_2017-11-02-13-13-27](https://user-images.githubusercontent.com/29614385/32377354-bb1e38bc-c07d-11e7-9df5-1604d84ac192.png)

This lag of drawing map on the screen (somewhat) demonstrates one of the observations I make earlier in the thread, where driving at night, the screen will often flash bright white as it loads new segments of the map, but CPU load is so high, it loads them in "day" mode before "slowly re-drawing them" in night mode (which is the darker theme).

**Comment by sonora** [comment]
The day/night mode observation is a separate issue and should not matter here. I also occasionally observe this, as follows: Normally, when the map is panned, the background default color should be that of the mode in which the map is displayed (at least ever since v2.8.x or so), e.g. dark green for night mode, something grey/whitish for day mode. It seems that when sometimes OsmAnd switches from an app profile which uses day mode to one which uses night mode (or maybe also when the day/night mode switches based on sunrise/sunset, not sure), the panning of the night view map still appears on the gray/whitish background, producing the "flashing" appearance. Killing/re-starting the app at this time fixes it. But this issue should not be the prime root cause for the CPU usage here.

**Comment by tazva** [comment]
I just presumed it may have been an offshoot, because it will sometimes take 2..5..12 seconds for the map to colorize properly, as I am travelling. In other words if I am travelling south, the bottom half of the screen will be tiles of day mode, which eventually turn dark. I just presumed the high CPU load was causing it to "take awhile".

Your description of it matches mine, it is a very good description.

**Comment by sonora** [comment]
To get back to the original issue: Are we quite confident the high CPU usage is
- (1) connected to when the device is or was in motion (i.e. does not occur when the device is at rest from the start)
- (2) connected to the street name display being enabled or disabled?

**Comment by tazva** [comment]
On point (1), It absolutely occurs regardless of motion--while completely at rest.

Here's a series of tests I ran lastnight using the same methodology as before. I only plotted load this time, no speed (no motion..). Device was powered on and left abandoned for periods of time. Capture intervals for data points was 10sec. I clipped out the first 30 seconds of boot time, as that would prejudice the chart. What is seen below is the device simply at rest:

![osmandplus_load_motionless](https://user-images.githubusercontent.com/29614385/32502540-031f0582-c3a9-11e7-987f-751741034be0.jpg)

There are four tests, all of them the same except for test 3. For this particular test, I blocked the GPS from being able to receive a signal. This shows OSM doing something, even when no GPS data is being parsed. 30% load with 0/0 for satellite signal.

As to point (2), I can run the same test with the street display disabled.

**Comment by sonora** [comment]
Very interesting. So it seems what you see is not related to map rendering, as the map remains stationary. We must be doing something else continuously using lots of CPU.

Not sure I can reproduce the observation though: On Samsung devices, and using Samsung's built-in "Active apps" monitor, OsmAnd's CPI usage seems to strictly drop to 0.00% when OsmAnd is in the background (and the app monitor is in the foreground): When in motion after latest 3 seconds from when switching from OsmAnd to the "Active apps" app.  And when performing that switch while at rest, the Active app monitor immediately shows OsmAnd using 0.00% CPU, which seems to indicate that is also the value for when OsmAnd is in the foreground.

So I cannot actually confirm the observation OsmAnd is using any CPU at all, unless
- you are in motion and the map needs to get panned and re-rendered
- you are in motion and in an app profile with street name display active (like our car profile by default), which seems to add extra and excessive CPU load (we currently refresh the street name lookup rather frequently every 10m of position change or so)

**Comment by tazva** [comment]
This is a Samsung tablet I'm using, as I've stated before. I can also reproduce it on a Samsung Note 4 Edge, but I used the third party app to provide objective raw usage of hardware, which could then be written to a log file. I have no idea how Samsung's applet work.

Now, as promised, here's my data from point (2), where I disabled street display.

![osmandplus_load_motionless_streetless](https://user-images.githubusercontent.com/29614385/32625743-193e33ac-c55b-11e7-90c0-bf90343cc061.jpg)

Completely motionless with street displays disabled, there appears to be no difference with the previous test (street displays enabled).

Combining these results, with the previous--particularly noting the CPU usage with GPS signal blocked--as an outside analyst, I can only conclude something in the code is burning up CPU slices in a loop, and it only gets worse when there's GPS data stream involved. That it grows significantly with higher speeds (thus, more GPS data/longer distances between points), only adds to the theory.

Again, baseline study of the device without OSM+ loaded, shows 0-4% CPU usage at most--typically 0-1%. The other 20-40% is coming from OSM+, for whatever reason. I've run a few other apps, including Nav apps, just to test. They don't appear to have the strain/drain OSM+ does. I will examine them in more detail.

**Comment by CJTmmr** [comment]
I have seen the high CPU-use as well on Samsung Galaxy Camera 2 (in latest nightlies)
In my experience: after a fresh start the CPU-use stays high for a long time, even when OsmAnd is no longer on foreground.
The Street name and Lane marker are not shown in follow-me-mode.
But even when the map is on a fixed position the CPU-use stays high.
I guess the disappearing streetname is a consequence of the high CPU-use, rather than a cause.

Only after a search for address or POI is completed, the CPU-use goes back to normal...
Is it possible that at start-up some initial search-proces is started, which somehow does not end?

Good luck investigating.

**Comment by tazva** [comment]
Thanks for that input CJ, I will try the POI/address search you suggested, to see if that does something.

**Comment by scaidermern** [comment]
To the developers: Please just use a profiler. It shouldn't be too hard to identify the cause of this high CPU consumption. Fixing this bug will help a lot of your users and improve one of the main point of criticism of OsmAnd (= consuming lots of CPU / battery).

**Comment by tazva** [comment]
Rounding the corner on almost a year old, I wonder how the progress is? I haven't had a chance to do any additional debugging since my last post, I've been very busy :(

I can no longer reliably use OSMand for navigation purposes due to the CPU load. A couple of times in the past month I've needed to navigate somewhere and would input the location. Inevitably I'd choose a destination, and it would take ages to calculate. Eventually it worked, but I would have to change route due to some issues, and recalculation simply took forever--and every intersection was a recalculation, so this forced it to just keep lagging and lagging. Ended up just killing it off and toggling over to google maps, it was up and running to my nav point within 3 seconds, with no issue.

I simply use OSMand for general quick-visual reference now, the CPU overhead is simply too great to do anything else with it. It still burns a hole in my hand if I use it on a more modern device, where the battery drains as quickly as pouring water from a glass.

**Comment by vshcherb** [comment]
I guess here there were measurements about navigation mode without calculating route, so in that case "reverse geocoding" is happening every 50 meters which is very CPU expensive for each call. Unfortunately no progress was made in that direction

**Comment by CJTmmr** [comment]
CPU load is high sometimes, even while navigating along a calculated route. And even while standing still, even when OsmAnd is running in background..... Unfortunately

**Comment by tazva** [comment]
@vshcherb as @CJTmmr  and some others have pointed out, it's not just in one mode, it's in every mode, whether calculating or not, whether in navigation mode, or in exploration mode. Something is clearly going on. Seeing all of the new bugs popping up, all in some way relating to 'slowdown', on brand-new multi-core very fast devices... this is pointing to something eating up processor time.

I can play HD games on my new phone and use less battery than OSM will just driving around. Yes, it is unfortunate no progress has been made to route this out, as mentioned last year, this will continue to plague OSM until fixed.

This issue was formerly tagged as a milestone event, but appears to no longer be the case. Rolling out new features is great, but shouldn't be a focus until foundational problems are resolved. Not sure why this isn't in the top 3 priorities. Just speaking from an outsider's perspective, of course.

**Comment by tazva** [comment]
While it's great to see OSMAnd is at a whole new version, and all sorts of features are being added, it would be great to fix underlying issues. Like #3898.
